### PR TITLE
KT-34715 False positive intention "Replace with ordinary assignment" when augmented assignment resolved to operator function

### PR DESCRIPTION
**Pull request by t-kameyama**
https://youtrack.jetbrains.com/issue/KT-34715

### CODE REVIEW

File: idea/src/org/jetbrains/kotlin/idea/intentions/ReplaceWithOrdinaryAssignmentIntention.kt [hunk 1]
```diff
@@ -8,19 +8,28 @@ package org.jetbrains.kotlin.idea.intentions
...
 import org.jetbrains.kotlin.psi.createExpressionByPattern
+import org.jetbrains.kotlin.resolve.calls.callUtil.getType
+import org.jetbrains.kotlin.resolve.lazy.BodyResolveMode
 
 class ReplaceWithOrdinaryAssignmentIntention :
     SelfTargetingIntention<KtBinaryExpression>(KtBinaryExpression::class.java, KotlinBundle.message("replace.with.ordinary.assignment")),
     LowPriorityAction {
     override fun isApplicableTo(element: KtBinaryExpression, caretOffset: Int): Boolean {
         if (element.operationToken !in KtTokens.AUGMENTED_ASSIGNMENTS) return false
-        if (element.left !is KtNameReferenceExpression) return false
```
Comment (nikitabobko) [inline idea/src/org/jetbrains/kotlin/idea/intentions/ReplaceWithOrdinaryAssignmentIntention.kt:22]: Code which was there before and what you're writing now are both absolutely wrong. I changed issue title: https://youtrack.jetbrains.com/issue/KT-34715 and wrote comment in issue.

Right implementation does this:
1. Resolve augmented assignment
2. Check that it resolves to `plusAssign`, `minusAssign`, etc
3. If so then intention shouldn't be suggested

E.g. this intention should work for such cases (but it doesn't work for such cases even now):
```
class Foo {
    var x: Int = 1
        get() {
            println("getter")
            return field
        }
        set(value) {
            println("setter")
            field = value
        }
}

fun main() {
    val foo = Foo()
    foo.x += 1
}
```

Comment (nikitabobko) [reply idea/src/org/jetbrains/kotlin/idea/intentions/ReplaceWithOrdinaryAssignmentIntention.kt:22]: So will you continue working on this PR? I can fix this issue

**Comment by nikitabobko** [comment]
Your PR breaks such tests:
`QuickFixTestGenerated$SuspiciousCollectionReassignment$ReplaceWithOrdinaryAssignment.testSimple`
`org.jetbrains.kotlin.checkers.JavaAgainstKotlinBinariesCheckerTestGenerated#testJvmOverloadsFunctions`

**Comment by nikitabobko** [comment]
Ah, wait it seems that I'm wrong about `org.jetbrains.kotlin.checkers.JavaAgainstKotlinBinariesCheckerTestGenerated#testJvmOverloadsFunctions`.

But I'm sure about `QuickFixTestGenerated$SuspiciousCollectionReassignment$ReplaceWithOrdinaryAssignment.testSimple`

**Comment by t-kameyama** [comment]
I've fixed in fa89ce0.

**Comment by yanex** [comment]
Review: OK. Additional patch:

```
Index: idea/src/org/jetbrains/kotlin/idea/intentions/ReplaceWithOrdinaryAssignmentIntention.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- idea/src/org/jetbrains/kotlin/idea/intentions/ReplaceWithOrdinaryAssignmentIntention.kt	(revision e19c2b31637f9dbc936f2ef96c663bf5265aca25)
+++ idea/src/org/jetbrains/kotlin/idea/intentions/ReplaceWithOrdinaryAssignmentIntention.kt	(date 1601990890878)
@@ -8,9 +8,11 @@
 import com.intellij.codeInsight.intention.LowPriorityAction
 import com.intellij.openapi.editor.Editor
 import org.jetbrains.kotlin.idea.KotlinBundle
-import org.jetbrains.kotlin.idea.references.mainReference
+import org.jetbrains.kotlin.idea.caches.resolve.resolveToCall
 import org.jetbrains.kotlin.lexer.KtTokens
 import org.jetbrains.kotlin.psi.*
+import org.jetbrains.kotlin.resolve.lazy.BodyResolveMode
+import org.jetbrains.kotlin.util.OperatorNameConventions
 import org.jetbrains.kotlin.utils.addToStdlib.safeAs
 
 class ReplaceWithOrdinaryAssignmentIntention : SelfTargetingIntention<KtBinaryExpression>(
@@ -24,7 +26,9 @@
         val left = element.left ?: return false
         if ((left.safeAs<KtQualifiedExpression>()?.selectorExpression ?: left) !is KtNameReferenceExpression) return false
         if (element.right == null) return false
-        return operationReference.mainReference.resolve().safeAs<KtFunction>()?.name?.endsWith("Assign") != true
+
+        val resultingDescriptor = operationReference.resolveToCall(BodyResolveMode.PARTIAL)?.resultingDescriptor ?: return false
+        return resultingDescriptor.name !in OperatorNameConventions.ASSIGNMENT_OPERATIONS
     }
 
     override fun applyTo(element: KtBinaryExpression, editor: Editor?) {
```

**Comment by nikitabobko** [comment]
Additional patch:
```
diff --git a/kotlin/idea/src/org/jetbrains/kotlin/idea/intentions/ReplaceWithOrdinaryAssignmentIntention.kt b/kotlin/idea/src/org/jetbrains/kotlin/idea/intentions/ReplaceWithOrdinaryAssignmentIntention.kt
index 3931b4389ea..d3173896444 100644
--- a/kotlin/idea/src/org/jetbrains/kotlin/idea/intentions/ReplaceWithOrdinaryAssignmentIntention.kt
+++ b/kotlin/idea/src/org/jetbrains/kotlin/idea/intentions/ReplaceWithOrdinaryAssignmentIntention.kt
@@ -22,7 +22,7 @@ class ReplaceWithOrdinaryAssignmentIntention : SelfTargetingIntention<KtBinaryEx
         if (!operationReference.textRange.containsOffset(caretOffset)) return false
         if (element.operationToken !in KtTokens.AUGMENTED_ASSIGNMENTS) return false
         val left = element.left ?: return false
-        if ((left.safeAs<KtQualifiedExpression>()?.selectorExpression ?: left) !is KtNameReferenceExpression) return false
+        if (left.safeAs<KtQualifiedExpression>()?.receiverExpression is KtQualifiedExpression) return false
         if (element.right == null) return false
         return operationReference.mainReference.resolve().safeAs<KtFunction>()?.name?.endsWith("Assign") != true
     }
diff --git a/kotlin/idea/test/org/jetbrains/kotlin/idea/intentions/IntentionTestGenerated.java b/kotlin/idea/test/org/jetbrains/kotlin/idea/intentions/IntentionTestGenerated.java
index b484f4837f6..ab57a6adba0 100644
--- a/kotlin/idea/test/org/jetbrains/kotlin/idea/intentions/IntentionTestGenerated.java
+++ b/kotlin/idea/test/org/jetbrains/kotlin/idea/intentions/IntentionTestGenerated.java
@@ -14884,6 +14884,11 @@ public abstract class IntentionTestGenerated extends AbstractIntentionTest {
             runTest("testData/intentions/replaceWithOrdinaryAssignment/complexRightExpression.kt");
         }
 
+        @TestMetadata("doubleQualifiedExpressionIsntAllowed.kt")
+        public void testDoubleQualifiedExpressionIsntAllowed() throws Exception {
+            runTest("testData/intentions/replaceWithOrdinaryAssignment/doubleQualifiedExpressionIsntAllowed.kt");
+        }
+
         @TestMetadata("minusAssignOperatorFun.kt")
         public void testMinusAssignOperatorFun() throws Exception {
             runTest("testData/intentions/replaceWithOrdinaryAssignment/minusAssignOperatorFun.kt");
diff --git a/kotlin/idea/testData/intentions/replaceWithOrdinaryAssignment/doubleQualifiedExpressionIsntAllowed.kt b/kotlin/idea/testData/intentions/replaceWithOrdinaryAssignment/doubleQualifiedExpressionIsntAllowed.kt
new file mode 100644
index 00000000000..59488d89bad
--- /dev/null
+++ b/kotlin/idea/testData/intentions/replaceWithOrdinaryAssignment/doubleQualifiedExpressionIsntAllowed.kt
@@ -0,0 +1,15 @@
+// IS_APPLICABLE: false
+class Bar {
+    var x: Foo = Foo()
+}
+
+class Foo {
+    var x: Int = 1
+}
+
+fun main() {
+    val foo = Bar()
+    // not allowed because it'd change semantic because
+    // of side effects of getters and setters
+    foo.x.x <caret>+= 1
+}
```

**Comment by nikitabobko** [comment]
Merged to new repo:
https://github.com/JetBrains/intellij-kotlin/commit/2063242394025480d50b99bea6f6a12e0a7f4b73
https://github.com/JetBrains/intellij-kotlin/commit/1863c65c759c9ee20f8456a74d0e71fd9c0e664e

**Pull request closed without merging**